Improve performance of migration verification

Lightwood13 opened the pull request:
This pull request fixes the "Verify migration" task hanging by adding some default interface methods from SchemaCrawler interfaces to ObjectDiffer exceptions. As explained more throughly in [this comment](https://github.com/sqldelight/sqldelight/issues/4759#issuecomment-3692948019), these methods create multiple paths to the same objects in the object graph, so the ObjectDiffer ends up comparing the same objects multiple times, leading to an exponential time complexity.

griffio commented:
> I wonder if we can find a way to prevent regression in the future.

I can do this in another PR 

One idea for a test is to add in another gradle MigrationTest based on the test case schema that has at least 3-4 interconnected tables and foreign keys and use `measureTimeMillis` to verify build time within say 30 secs  - https://github.com/sqldelight/sqldelight/blob/5d9afd5789464791e31ee4b5dbf9c28b356b296b/sqldelight-gradle-plugin/src/test/kotlin/app/cash/sqldelight/tests/MigrationTest.kt#L346

GitHub CI is currently running tests under `JDK Zulu 17.62.17` this contains the`interface default method` fix that causes the whole issue with an interconnected graph of multiple paths to the same object.
e.g 
```
 @Test
  fun `verifySqlDelightMigration task completes in a reasonable time`() {
    val runner = GradleRunner.create()
      .withCommonConfiguration(File("src/test/migration-performance-regression"))
      .withArguments("clean", "verifyMainDatabaseMigration", "--stacktrace")

    val duration = measureTimeMillis {
      val result = runner.build()
      assertThat(result.output).contains("BUILD SUCCESSFUL")
    }

    assertThat(duration)
      .withFailMessage("verifyMigration task took too long: ${duration}ms. Potential complexity regression.")
      .isLessThan(30_000)
  }
  ```
   
Also - a smaller perf issue is the regex in https://github.com/sqldelight/sqldelight/blob/5d9afd5789464791e31ee4b5dbf9c28b356b296b/sqlite-migrations/src/main/kotlin/app/cash/sqlite/migrations/ObjectDifferDatabaseComparator.kt#L61 is being called 💯 of times even in the small testcase - these could be precompiled once

griffio commented:
> Maybe there's a way to count the number of comparisons in the ObjectDifferDatabaseComparator?

Yes, counting something is much more deterministic than timings. 
I had already tried the gradle runner, with just two table relationships and debug output `./gradlew verifySqlDelightMigration --debug` to try and count log accessor lines.
The debug output log is large (e.g 48MB) and bigger (e.g 68MB) in slow-case, still not a huge difference.
Currently - with the test case 3 or 4 tables, the timings are around 12 secs with fix and over 1 min without. 
The problem is that even with the fix, the simplest schema test cases are quite slow and using a real schema it may still take several minutes. 
🔬  I will continue to look into.

Lightwood13 commented:
We can count comparisons by adding a custom Differ. Here's a quick implementation for testing where I output number of comparisons as error:
[Count_comparisons.patch](https://github.com/user-attachments/files/24355012/Count_comparisons.patch)
I've got 4210345 comparisons without the fix and 793467 with the fix. This still seems like quite a lot for such a small schema, there are probably more things to optimize.

brinsche commented:
Isn't `comparison().ofType(Class.forName("schemacrawler.crawl.MutableForeignKey")).toUseEqualsMethod()` the correct fix then or is it still expensive or incorrect somehow?

Lightwood13 commented:
I think it's incorrect as `MutableForeignKey` doesn't override the `equals` method. It will use the `equals` methods of superclasses. The only superclass with equals method is `AbstractDependantObject`, which just compares parent table, ignoring other fields of `MutableForeignKey`.

griffio commented:
> I think it's incorrect as `MutableForeignKey` doesn't override the `equals` method. It will use the `equals` methods of superclasses. The only superclass with equals method is `AbstractDependantObject`, which just compares parent table, ignoring other fields of `MutableForeignKey`.

I looked at the last version of `SchemaCrawlerDifferBuilder` from where SqlDelight took the initial code years ago.  https://github.com/schemacrawler/SchemaCrawler/blob/v16.21.4/schemacrawler-diff/src/main/java/schemacrawler/tools/integration/objectdiffer/SchemaCrawlerDifferBuilder.java before it was removed from (schemacrawler) based on this comment https://github.com/schemacrawler/SchemaCrawler/issues/1293#issuecomment-1769719566

As you mentioned above ⬆️  - this adds the `primaryKeyTable` and `foreignKeyTable` to the exclusions and significantly speeds up my worst case example to run in seconds. Using the counter patch `Performed 51876 comparisons` and compared to taking many minutes with the current fixes.
e.g 🔥 
```
-- Table 1: The Users
CREATE TABLE users (
  user_id INTEGER NOT NULL PRIMARY KEY,
  username TEXT NOT NULL
);

-- Table 2: Points to Users
CREATE TABLE profiles (
  profile_id INTEGER NOT NULL PRIMARY KEY,
  user_id INTEGER NOT NULL,
  FOREIGN KEY (user_id) REFERENCES users(user_id)
);

-- Table 3: Points to Users and Profiles (Two paths to reach 'users')
CREATE TABLE posts (
  post_id INTEGER NOT NULL PRIMARY KEY,
  user_id INTEGER NOT NULL,
  profile_id INTEGER NOT NULL,
  FOREIGN KEY (user_id) REFERENCES users(user_id),
  FOREIGN KEY (profile_id) REFERENCES profiles(profile_id)
);

-- Table 4: Points to Profiles and Posts (Multiple paths to 'users' and 'profiles')
CREATE TABLE comments (
  comment_id INTEGER NOT NULL PRIMARY KEY,
  profile_id INTEGER NOT NULL,
  post_id INTEGER NOT NULL,
  FOREIGN KEY (profile_id) REFERENCES profiles(profile_id),
  FOREIGN KEY (post_id) REFERENCES posts(post_id)
);

-- Table 5: The Multi-Liker (Creates an explosion of paths)
CREATE TABLE likes (
  like_id INTEGER NOT NULL PRIMARY KEY,
  user_id INTEGER NOT NULL,
  post_id INTEGER NOT NULL,
  comment_id INTEGER NOT NULL,
  FOREIGN KEY (user_id) REFERENCES users(user_id),
  FOREIGN KEY (post_id) REFERENCES posts(post_id),
  FOREIGN KEY (comment_id) REFERENCES comments(comment_id)
);
```

Lightwood13 commented:
Checked, `MutableForeignKey` has a `getColumnReferences` getter which returns an instance of `ColumnReference` which contains `getPrimaryKeyColumn` and `getForeignKeyColumn`. Those columns have names of their corresponding tables in the `fullName` property. This means that we can safely add properties `foreignKeyTable` and `primaryKeyTable` of `MutableForeignKey` to ObjectDiffer exceptions as discussed before, as names of the tables related to foreign key are compared through `getColumnReferences`.

I will continue looking into what happens during schema comparison, as 50k comparisons still seems like too many for such a simple schema.

dees91 commented:
I tested this fix on my SQLite schema with 48 tables and large number of foreign key relationships.

I measured the execution time of the comparison itself:

```kotlin
override fun compare(db1: CatalogDatabase, db2: CatalogDatabase): DatabaseDiff {
    val diff = measureTimedValue {
        ObjectDifferDatabaseDiff(
            differBuilder().compare(db1.catalog, db2.catalog),
            db1,
            db2,
        )
    }
    println("ObjectDifferDatabaseDiff duration: ${diff.duration}")
    return diff.value
}
```

The current PR excludes alone do not reduce the time enough for my schema - it still takes over 10 minutes. However, the additional excludes and/or [toUseEqualsMethod()](https://github.com/sqldelight/sqldelight/pull/6073#issuecomment-3694393793) discussed here bring the time down to seconds:

| Property excludes | `toUseEqualsMethod()` | Duration |
|---|---|---|
| `referencedTable`, `referencedTables`, `referencedObjects`, `dependentTable`, `dependentTables` | - | **> 10 min** |
| `referencedTable`, `referencedTables`, `referencedObjects`, `dependentTable`, `dependentTables`, `foreignKeyTable`, `primaryKeyTable` | - | **14.77s** |
| `referencedTable`, `referencedTables`, `referencedObjects`, `dependentTable`, `dependentTables`, `foreignKeyTable`, `primaryKeyTable`, `referencingTables` | - | **4.81s** |
| `referencedTable`, `referencedTables`, `referencedObjects`, `dependentTable`, `dependentTables` | `MutableForeignKey` | **2.16s** |
| `referencedTable`, `referencedTables`, `referencedObjects`, `dependentTable`, `dependentTables`, `foreignKeyTable`, `primaryKeyTable`, `referencingTables` | `MutableForeignKey` | **632.93ms** |

Lightwood13 commented:
I did some more debugging and and as a result found a few more properties that can be excluded, getting the number of comparisons down from 53270 to 8222. I've added those excluded properties, comparison count and a test to this PR. Here's a more detailed explanation of all the changes.

I replaced `propertyName` to `propertyNameOfType` when adding excluded properties. It matches only properties of given type instead of all properties with given name. This way we won't accidentally ignore properties with the same name in other classes and it will be clear for future maintainers what specific method we wanted to exclude, so they can find it easily.

Here's the full list of new excluded properties in this pull request with reason why we can safely exclude them:
- Class `MutableTable`:
    - `getReferencingTables()`, `getReferencedTables()` from interface `Table` - no need to compare here, they will be compared directly through foreign keys.
- Class `MutableForeignKey`:
    - `getReferencingTable()`, `getReferencedTable()`, `getPrimaryKeyTable()`, `getForeignKeyTable()` from interface `TableReference` - no need to compare whole tables, we just need to compare table names, which will be done when comparing columns through `getColumnReferences().getPrimaryKeyColumn()` and `getColumnReferences().getForeignKeyColumn()`.
- Interface `Column` (its methods are excluded using just property name without type, as there are multiple classes implementing this interface, so I think it’s simpler to just exclude by property name):
    - `getReferencedColumn()` - no need to compare it here, it will be compared when comparing foreign keys.
- Interface `BaseColumn` (also excluded using just property name):
    - `getColumnDataType()` - a duplicate of `type()` from interface `TypedObject`, which `BaseColumn` extends.

Next, I've added a test to check number of comparisons and prevent future regressions. It uses DB schema provided by @griffio, but I had to remove one of the tables, otherwise the test wouldn't complete in reasonable time when testing on codebase before the fix.

In initial version of this PR, there were a few excluded properties from a newer version of `SchemaCrawler` than the version, that is used currently. I did it to prevent regressions when `SchemaCrawler` version is bumped. But, as there is a test now to monitor for regressions, we don't need those properties anymore, so I removed them.

The final performance difference in terms of number of comparisons before and after the fix:
- for schema provided by @brinsche: 4210345 -> 8222 (x512)
- for schema used in tests (same as schema provided by @griffio without the `likes` table): 2273863 -> 6780 (x335)
- with the `likes` table the speedup is even bigger, because there number of comparisons gets larger exponentially when number of foreign keys increases. I got around 10k comparison with the fix, and without the fix the test run too long without finishing. With some of the properties excluded there were 17m comparisons, so speedup here is greater than x17000.

There are still some optimizations possible, for example, by changing comparison of `MutableColumnDataType` to be compared by just name instead of all of its 20 fields I got another 1.7x speedup, but I'm not sure if it's safe to do comparison like this. Anyway, I think the current implementation should be good enough for now, as we removed the slowdown from comparing tables multiple times through foreign keys. This is the most important slowdown because it gets exponentially slower when the number of foreign keys increases.

dees91 commented:
> > I tested this fix on my SQLite schema with 48 tables and large number of foreign key relationships.
> 
> @dees91 thank you for testing the fix on your schema. Can you please also test the latest commit?

These are 5 measurements for the latest version of `ObjectDifferDatabaseComparator` from this PR:
```
ObjectDifferDatabaseDiff duration: 719.284500ms
ObjectDifferDatabaseDiff duration: 746.570792ms
ObjectDifferDatabaseDiff duration: 730.560875ms
ObjectDifferDatabaseDiff duration: 608.446208ms
ObjectDifferDatabaseDiff duration: 747.771875ms
```

It looks like really nice performance 🙂

griffio commented:
@Lightwood13 thanks and lastly add a change log entry into a new fixed section in CHANGELOG.md
 
I can approve and 🧜‍♂️ merge, no doubt many will celebrate this PR that fixes #4759
